text/3617-precise-capturing.md

```diff
@@ -0,0 +1,668 @@
+
+```rust
+trait Trait {
+    type Foo<'a>: Sized where Self: 'a;
+    fn foo(&self) -> Self::Foo<'_>;
+}
+```
+
+...then an impl of this trait can provide a type for the associated type `Foo` that uses the `&'_ self` lifetime:
+
+```rust
+struct A;
+impl Trait for A {
+    type Foo<'a> = &'a Self; // Or, e.g.: `impl use<'a> Sized`
+    fn foo(&self) -> Self::Foo<'_> { self }
+}
+```
+
+However, such an impl may also provide a type that does *not* use the lifetime:
+
+```rust
+struct B;
+impl Trait for B {
+    type Foo<'a> = (); // Or, e.g.: `impl use<> Sized`
+    fn foo(&self) -> Self::Foo<'_> {}
+}
+```
+
+If we only know that the value is of some type that implements the trait, then we must assume that the type returned by `foo` *might* have used the lifetime:
+
+```rust
+fn test_trait<T: Trait + 'static>(x: T) -> impl Sized + 'static {
+    x.foo()
+//~^ ERROR cannot return value referencing function parameter `x`
+}
+```
+
+However, if we know we have a value of type `B`, we can *rely* on the fact that the lifetime was not used:
+
+```rust
+fn test_b(x: B) -> impl Sized + 'static {
+    x.foo() //~ OK.
+}
+```
+
+We would say that the impl for `B` is *refining* in that it offers more to or demands less of callers than the minimum the trait could offer or the maximum it could demand.  Associated type definitions are always refining in this way.
+
+RPITIT desugars into associated types similar to those above, but here we've currently decided to lint against this refinement, e.g.:
+
+```rust
+trait Trait {
+    fn foo(&self) -> impl Sized;
+}
+
+impl Trait for () {
+    fn foo(&self) -> () {}
+//~^ WARN impl trait in impl method signature does not match
+//~|      trait method signature
+//~| NOTE add `#[allow(refining_impl_trait)]` if it is intended
+//~|      for this to be part of the public API of this crate
+//~| NOTE we are soliciting feedback, see issue #121718
+//~|      <https://github.com/rust-lang/rust/issues/121718>
+//~|      for more information
+}
+```
+
+Similarly, for consistency, we'll lint against RPITIT cases where less is captured by RPIT in the impl as compared with the trait definition when using `use<..>`.  E.g.:
+
+```rust
+trait Trait {
+    fn foo(&self) -> impl Sized;
+}
+
+impl Trait for () {
+    fn foo(&self) -> impl use<> Sized {}
+//~^ WARN impl trait in impl method signature does not match
+//~|      trait method signature
+//~| NOTE add `#[allow(refining_impl_trait)]` if it is intended
+//~|      for this to be part of the public API of this crate
+//~| NOTE we are soliciting feedback, see issue #121718
+//~|      <https://github.com/rust-lang/rust/issues/121718>
+//~|      for more information
+}
+```
+
+## Argument position impl Trait
+
+Note that for a generic type parameter to be captured with `use<..>` it must have a name.  Anonymous generic type parameters introduced with argument position `impl Trait` (APIT) syntax don't have names, and so cannot be captured with `use<..>`.  E.g.:
```

Comment:
>cannot be captured with `use<..>`

Are they captured implicitly now?
If they are, then what is the opt out of that?

Are unnamed lifetimes implicitly captured by RPIT?
What is the opt out in that case?
(If unnamed early bound lifetimes are possible at all.)

Reply:
>Nope, we don't capture APITs or elided lifetimes with `use<'a, T>`.

My first question was whether they are captured if `use` is not written.
The text says that all generic parameters are captured (on 2024 edition at least), but it's not clear whether APITs and elided lifetimes are included into this "all".

Reply:
All type and const generic parameters (including APITs) are captured for all RPITs regardless if they're in [edition 2021](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9eb4c7182b8ad8979e08e6cdba89530e) or 2024.

The only change between 2021 and 2024 involves capturing all lifetimes parameters in scope (incl. elided lifetimes, since those become early- or late-bound lifetime params) -- previously we only captured lifetimes mentioned in the bounds of RPITs. The opt-out here is to use `use<'a, T>` with the set of lifetimes that previously showed up in your opaque's bounds.

The only corner case is when you have an APIT *and* a lifetime you want to not capture -- you'll need to turn your APITs into real type parameters.

Regarding opt-out for capturing type or const generic parameters currently, it's not possible [to do currently](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=843eed31837eb4ba6169df4a28dced86); we will likely support that eventually, but if you see https://github.com/rust-lang/rfcs/blob/TC/precise-capturing/text/3617-precise-capturing.md#stabilization-strategy, we're probably not going to stabilize that initially because it requires exercising new paths of the type system (namely, bivariant unconstrained type parameters).